README.md
# Support tekmunkey via Patreon

If you'd like to support my free software developments, my patreon can be found at:

https://www.patreon.com/tekmunkey

# ubuntuMUSH

The documentation on this product is best viewed inside the scripts themselves, though it is a bit sparse.

Create a directory such as ~/MUSH

Put this repo in that directory along with your .tar.gz distro file(s).  Make sure whatever MUSH/MUX platforms you downloaded from github or FTP sites match the filenames in these scripts (as defined on line 34 of each script, after the = and between double quotes).  If they don't match, then change the value between the double quotes to match the filename.

At the commandline:

run:  chmod ug+x ubuConfig.sh

run:  ./ubuConfig.sh

The script should install all needed packages and set the other scripts +x

run:  chmod ug+x ubuConfig.sh

run:  ./installPennMUSH.sh

Rename the myPenn-186 directory to whatever you want it to be.

Edit ~/doBackups.sh - READ THE DOCUMENTATION IN THAT FILE CAREFULLY - and add your ~/MUSH and maybe your ~/.ssh directories to the backupTargets array.

run:  cron -e

Add ~/doBackups.sh to a weekly backup (or however often you want to back up).

Add @reboot ~/MUSH/myMUSHDir/game/restart anywhere on its own line to start your MUSH when the server starts up

That's it!

At some point I may have time to go in and document each package installation but honestly it's pretty murky water.  PennMUSH's documentation really sucks.  They vaguely hint at what's required, their hints are outdated where they aren't outright wrong or completely missing, and several of the installed packages were determined by troubleshooting actual compile-time errors as opposed to translating Penn documentation into modernized Ubuntu 16 package names.  By the time I had figured out what package was required, in some cases, I had already forgotten exactly why.



CHANGELOG.md
# 2017-12-25

* Changelog initiated

* After troubleshooting a Ubu16 install on DreamHost, troubleshooting notes were added in the script comments.  See also TROUBLESHOOTING.md

* After troubleshooting some user-error issues, script functions were commented out and commentation was added for usage by decommentation, if practical conventions are observed (see comments at the bottom of the script file, referring to "enable the next step" operations).

# 2017-12-28

* Added some lib32gcc compiler support installs to ubuConfig script.  Helps with 64-bit operating systems.

* Added some /etc/skel mods to the script, including a by-default .ssh folder and a copy of the doBackups script for all newly created profiles 

# 2018-01-25

* Added installTinyMUX.sh to automate TinyMUX installation for Ubuntu systems with all the really useful options enabled

# 2018-03-10

* Converted MySQL/MariaDB installation to fully automated via debconf and CLI parameters.  User may pass -mrp=YourMySQLRootPassword or --mySQLRootPassword=YourMySQLRootPassword from the commandline to set up a custom root password for MySQL at runtime.  The default value is 'mysqlRoot' - the script will inform the user of exactly what the MySQL Root Password is when complete.

* Replaced mysql_secure_installation with bash/expect routines into the mysql CLI client, sending the relevant queries.  This eliminates automation-breaking user prompts.

* Added MediaWiki v1.30.0 installation w/prerequisites.  Verified functional.  After the script runs users should be able to browse to http://Server.IP.Address.Here/mediawiki 

# 2018-03-15

* Added muDBSchema, muDBUser, and muDBPass variables and corresponding CLI parameters for -mus=SchemaName, -muu=UserName, -mup=UserPassword which creates a MySQL Database and sets up a privileged User with the specified password.  The script will inform the user of exactly what these values are (Schema, Username, Password) when complete.

* Added randPass() function to script, which allows better/more secure default password values.  Default passwords for MySQL Root, MU DB User, and MediaWiki Admin are now randomly generated for each script run.  Values may still be overridden by CLI Parameters and the old (simplified) passwords are still there - you can always comment out the single line values that call on the randPass generator to keep the simple default passwords.

# 2018-03-21

* Added iptables/UFW install/configuration

* Fixed problem with getRandPass function

# 2018-03-22

* Removed dig installation which was preventing other network tools installs such as whois

* Moved sendmail to separate line and commented out (was preventing postfix installation)

doBackups.sh
#!/bin/bash
# chmod ugo+x your_shell_script.sh


# To compress:    tar -zcvf archive-name.tar.gz directory-name
# To uncompress:  tar -zxvf prog-1-jan-2005.tar.gz -C /tmp

#
# backupTargets is an array of directory names that will be backed up.  Each directory and its subdirectories will be 
# tarred and gzipped daily.
#
declare -a backupTargets=("$@")
## ie:  backupTargets+=("~/mydirectory")
## ie:  backupTargets+=("~/mydirectory/subdir")

#
# backupDirectory is the storage directory where all backup files will be dumped
# I recommend you put this script in your ~/ (home) directory or else 
# at the root of your directory where all your MUSH directories live, rather than 
# inside any particular MUSH directory.
#
# It's a good backup script script.
#
declare backupDirectory="./Backups"

#
# compilationFile is the backup compilation where all backup files are compiled, for a quick and easy offsite dump at a later time
# !!! DO NOT STORE THIS INSIDE THE backupDirectory !!!
#
declare compilationFile="./Backups.tar"

if [[ ! -d "${backupDirectory}" ]]; then
    mkdir "${backupDirectory}"
    chmod 700 "${backupDirectory}"
fi

if [[ ! -f "${compilationFile}" ]]; then
    touch "${compilationFile}"
    chmod 600 "${compilationFile}"
fi

#
# It's a good idea to back up your .ssh directory.  If you store this script at ~/ (your home)
# then you can just uncomment the next backupTargets line and it will automatically add 
# your .ssh dir to the backups.  If you store this script anywhere else you need to 
# modify the path value accordingly.  Just leave it inside parentheses because this is a bash 
# array value.
#
# backupTargets+=("./.ssh")
#
# The next backupTargets line should be uncommented and the path modified appropriately to target 
# the top level directory where your particular top level MUSH resident directory is.
#
# This should be the directory where, for example, your MUSH tar files are, your config files, 
# and whatever director/y/ies you untarred and compiled into.
#
# backupTargets+=("./your-mush-top-level-directory")
for bt in "${backupTargets[@]}"; do
    #
    # archFile is a timestamp followed by the directory name only of the entry from backupTargets that we're about to archive
    # This is a full chronological backup of the directory on that particular date
    #
    declare archFile="${backupDirectory}"/"$(date +%Y%m%d-%H%M).${bt##*/}.tar.gz"
    if [[ ! -f "${archFile}" ]]; then
        touch "${archFile}"
        chmod 600 "${archFile}"
    fi
    tar -zcpf "${archFile}" "${bt}"
    #
    # archComp is a compilation of all the archives for that particular directory, provided for a quick and easy offsite dump at a later time
    #
    declare archComp="${backupDirectory}"/"compilation-${bt##*/}.tar"
    if [[ ! -f "${archComp}" ]]; then
        touch "${archComp}"
        chmod 600 "${archComp}"
        tar -cpf "${archComp}" "${archFile}"
    else
        tar -puf "${archComp}" "${archFile}"        
    fi
done

tar -uf "${compilationFile}" "${backupDirectory}"

installTinyMUX.sh
#!/bin/bash
trap "exit 1" TERM
export SCRIPT_PID=$$ # use 'kill -s TERM $SCRIPT_PID' to pop this from anywhere in the script

######!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!######
#####                                                                                                    #####
#####       DO NOT RUN THIS SCRIPT UNLESS YOU ALREADY RAN AN OPERATING SYSTEM CONFIGURATION SCRIPT       #####
#####          THIS SCRIPT RELIES ON THE PRE-INSTALLATION AND CONFIGURATION OF REQUIRED PACKAGES         #####
#####                                                                                                    #####
######!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!######

#
# These variables SHOULD work for debian and ubuntu - ie: debian base OSes with MySQL/Maria ie: MysQL base DBs
#
declare dbConfig="/usr/bin/mysql_config"
declare dbInc="$(${dbConfig} --variable=pkgincludedir)"
declare dbLib="$(${dbConfig} --variable=pkglibdir)"

#
# Change the following value to whatever you want your game directory to be
#   When you uncomment a desired platform unpacker below, it will automagically unpack into this directory name
#   beneath whatever directory you're in now.
#   * Avoid spaces, backslashes, forwardslashes, question marks, and don't put a dot as the first character.
#     Generally speaking, know the rules and best practices for directory names in Linux
#   * DO NOT ADD ./ TO THIS STRING - the script automatically adds this value
#
#   Only change the part inside quotation marks :-)
#
declare gameDirectory="tinymux210"
#
# Change the following value to the filename of whatever package you downloaded.  This value is whatever immediately precedes the 
# .tar.gz in what you downloaded from the public/anonymous FTP Server at tinymux.org
#
declare packageName="mux-2.10.1.14.unix"

# The if statement is needed to ensure that the directory exists
# yes, Comic Book Guy, I could mkdir -p && tar -xf on one line below, but:
#   1:  At some future point I (or some other consumer) may want to add additional steps if/when the directory doesn't already exist
#   2:  Most folks who want to start up a MUSH are artists and writers, not geeks.
#       As mush platforms update and .tar filenames change, this script needs to be easily updateable and as non-intimidating as possible 
#       so I'm keeping eldritch runes and codey scripture to a minimum
if [[ ! -d "./${gameDirectory}" ]]; then
    mkdir "./${gameDirectory}"
    chmod 770 "./${gameDirectory}"
fi

# I assume that you downloaded the TinyMUX 2.10.1.14 archive from the tinymux.org public/anonymous ftp server
gunzip "./${packageName}.tar.gz"
# The next line untars our archive file
# the .tar is the file in question
# the -C and its following parameter is our target directory
# --strip-components=1 allows renaming of the "default" directory that the platform devs crammed in there when they tarballed it in the first place
tar -xf "./${packageName}.tar" -C "./${gameDirectory}" --strip-components=1
# move our working directory into the game directory
cd "./${gameDirectory}/src"
#
# Install TinyMUX with
#   * Reality Levels, WoD Realms, Inline AND Async SQL Enabled
#
./configure --enable-realitylvls --enable-wodrealms --enable-inlinesql --enable-stubslave --with-mysql-include="${dbInc}" --with-mysql-libs="${dbLib}"
make depend
make
cd ./modules
./configure
make
# return to where this script resides
cd ../../..

installPennMUSH.sh
#!/bin/bash
trap "exit 1" TERM
export SCRIPT_PID=$$ # use 'kill -s TERM $SCRIPT_PID' to pop this from anywhere in the script

######!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!######
#####                                                                                                    #####
#####       DO NOT RUN THIS SCRIPT UNLESS YOU ALREADY RAN AN OPERATING SYSTEM CONFIGURATION SCRIPT       #####
#####          THIS SCRIPT RELIES ON THE PRE-INSTALLATION AND CONFIGURATION OF REQUIRED PACKAGES         #####
#####                                                                                                    #####
######!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!######

#
# These variables SHOULD work for debian and ubuntu - ie: debian base OSes with MySQL/Maria ie: MysQL base DBs
#
declare dbConfig="/usr/bin/mysql_config"
declare dbInc="$(${dbConfig} --variable=pkgincludedir)"
declare dbLib="$(${dbConfig} --variable=pkglibdir)"

#
# Change the following value to whatever you want your game directory to be
#   When you uncomment a desired platform unpacker below, it will automagically unpack into this directory name
#   beneath whatever directory you're in now.
#   * Avoid spaces, backslashes, forwardslashes, question marks, and don't put a dot as the first character.
#     Generally speaking, know the rules and best practices for directory names in Linux
#   * DO NOT ADD ./ TO THIS STRING - the script automatically adds this value
#
#   Only change the part inside quotation marks :-)
#
declare gameDirectory="myPenn-186"
#
# Change the following value to the filename of whatever package you downloaded.  This value is whatever immediately precedes the 
# .tar.gz in what you downloaded from Penn's github.
#
declare packageName="pennmush-186p1"

# The if statement is needed to ensure that the directory exists
# yes, Comic Book Guy, I could mkdir -p && tar -xf on one line below, but:
#   1:  At some future point I (or some other consumer) may want to add additional steps if/when the directory doesn't already exist
#   2:  Most folks who want to start up a MUSH are artists and writers, not geeks.
#       As mush platforms update and .tar filenames change, this script needs to be easily updateable and as non-intimidating as possible 
#       so I'm keeping eldritch runes and codey scripture to a minimum
if [[ ! -d "./${gameDirectory}" ]]; then
    mkdir "./${gameDirectory}"
    chmod 770 "./${gameDirectory}"
fi

# I assume that you downloaded the PennMUSH 1.8.6 patchlevel 1 archive from github
gunzip "./${packageName}.tar.gz"
# The next line untars our archive file
# the .tar is the file in question
# the -C and its following parameter is our target directory
# --strip-components=1 allows renaming of the "default" directory that the platform devs crammed in there when they tarballed it in the first place
tar -xf "./${packageName}.tar" -C "./${gameDirectory}" --strip-components=1
# move our working directory into the game directory
cd "./${gameDirectory}"
./configure --without-sqlite3 --without-postgresql --with-mysql="${dbConfig}"
make update
make install
# return to where this script resides
cd ../

ubuConfig.sh
#!/bin/bash
trap "exit 1" TERM
export SCRIPT_PID=$$ # use 'kill -s TERM $SCRIPT_PID' to pop this from anywhere in the script

#
# You will probably need to run the following from the commandline before this script will run
#   * Do that for THIS SCRIPT ONLY!
#
# chmod 770 your_shell_script.sh
#

#####
# HOPEFULLY THIS IS THE FIRST THING YOU DO AFTER GETTING YOUR NEWLY HOSTED SERVER ONLINE, OTHERWISE IT MAY BORK SOME SETTINGS YOU SET UP
#####

#
# This particular script is tweaked for ubuntu 16 server.
# 
# BEFORE YOU BEGIN:
#   * Ensure that your network is fully and properly configured, especially your DNS.
#   * Pre-testing sudo apt-get update && sudo apt-get upgrade to see if there are any errors may be one way of doing this
#
#   * This script will pull everything you need to successfully compile PennMUSH v1.8.6rc1p1 with MySQL support enabled.
#   * This script will pull everything you need to have a lAMP (linux-based Apache MySQL Preprocessor server running
#     ** This script will install both PHP and Python as well as run the Apache Mod configurations needed to run both 
#        PHP and Python scripts as CGI Preprocessor pages.
#     ** Successfully configuring individual sites via sites-available, sites-enabled, .htaccess, and PHP or Python scripts 
#        is still up to you.
#

#
# The next line ensures that things like the mysql-server installation don't interrupt the script
# with annoying user-interactive selection menus
#
export DEBIAN_FRONTEND=noninteractive

#
# If at any time your machine gives you errors (typically on any sudo callout) "unable to resolve host <insert your machine name here>" 
# then you must edit your /etc/hosts file and add your machine name after localhost
#
# ALWAYS MAKE DAMN SURE YOU KEEP LOCALHOST AS AN ALIAS FOR 127.0.0.1
#
# The tippy-top first line of your /etc/hosts file should look something like:
# ie:  127.0.0.1 localhost myMachineName someAdditionalAlias
#

#
# If ubu16 sticks on apt-get update while trying to connect to security.ubuntu.com, then your server has a problem connecting 
# over IPv6.  This is an issue with your host or ISP so all you can really do is fix up your linux box.
#
# To do that you must:
#     sudo nano /etc/gai.conf
#   * (or whatever your favorite/installed editor is, if not nano)
#
# Under the line:  
#     # For sites which prefer IPv4 connections change the last line to
#
# Uncomment what follows by removing its # (don't remove the # in THIS file - remove the # in gai.conf)
# Change from:
#     # precedence ::ffff:0:0/96 100
# to:
#     precedence ::ffff:0:0/96 100
#
# This script will NOT attempt to auto-correct this problem for you because it does not generate any error code and that means 
# the script would have to run the "fix" up front and in all events, and it simply isn't needed in most cases.
#

#
# Run this with sudo, never as root
#
# Run this BEFORE you compile or install any MUSH/MUX platform
#

#
# This function generates a random password between $1 and $2 characters in length
#
function getRandPass
{
    #
    ## $1 must be the minimum number of characters requested
    ## $2 must be the maximum number of characters requested
    #

    local -i minLength=8
    #
    # When referencing this value in code, DO NOT enclose in single or double quotes
    #
    local isNumRegEx="^[0-9]+$"
    ## Validate that $1 is not null and is a number and is > 4 (an oldschool minimum password length)
    if [[ ! -z "${1}" ]] && [[ "${1}" =~ ${isNumRegEx} ]] && [ "${1}" -gt 4 ]; then
        minLength=${1}
    fi

    local -i maxLength=12
    #
    ## Validate that $2 is not null and is a number and is > 4 (an oldschool minimum password length)
    #
    if [[ ! -z "${2}" ]] && [[ "${2}" =~ ${isNumRegEx} ]] && [ "${2}" -gt 4 ]; then
        maxLength=${2}
    fi

    #
    ## Ensuring that minlength isn't > maxLength - if it is then reverse them
    ##   * DO NOT TRY TO PUT THIS ABOVE, INTO THE VALUE VALIDATIONS... IF ANY VALUE VALIDATION FAILS THEN EITHER MIN OR MAX LENGTH MAY BE DEFAULTED
    ##     THIS METHOD ABSOLUTELY ENSURES THAT getRandPass WILL ALWAYS RETURN A LEGITIMATELY RANDOM-LENGTH AND RANDOM-CHARS PASSWORD, EVEN IF IT THROWS OUT INVALID PARAMETERS
    #
    if [ "${minLength}" -gt "${maxLength}" ]; then
        local -i minBak=${minLength}
        minLength=${maxLength}
        maxLength=${minBak}
    fi
    
    #
    ## Get a random number between minLength and maxLength and that is the actual password length for output
    #
    local -i passLength=$(shuf --input-range="${minLength}-${maxLength}" --head-count=1)

    echo $(< /dev/urandom tr -dc '!@#$%^&*=+<>:_A-Z-a-z-0-9' | head --bytes="${passLength}";echo;)
}


#
# The mysqlRootPass value will be injected into the MySQL/MariaDB installation at runtime.
#
# To set this value when invoking the script from commandline, call:
#   scriptName.sh -mrp=YourPass
#     OR
#   scriptName.sh --mysqlRootPass=YourPass
#
declare mysqlRootPass="mysqlRoot"
#
# Comment out the next line if you want a simplified mysqlRoot password - both this and the initial/simple value above may be overridden with the CLI arguments
#
mysqlRootPass=$(getRandPass 12 16)

#
# The muDBSchema value will be injected into SQL queries at runtime to create a database schema for your MU
#
# To set this value when invoking the script from commandline, call:
#   scriptName.sh -mus=YourSchemaName
#     OR
#   scriptName.sh --muDBSchema=SchemaName
#
declare muDBSchema="muDB"
#
# The muDBUser value will be injected into a newly created user with full access to muDBSchema at runtime
#
# To set this value when invoking the script from commandline, call:
#   scriptName.sh -muu=UserNameFormuDBSchema
#     OR
#   scriptName.sh -muDBUser=UserNameFormuDBSchema
#
declare muDBUser="muDBUser"
#
# The muDBPass value will be injected into the newly created MySQL muDBUser as its password
#
# To set this value when invoking the script from commandline, call:
#   scriptName.sh -mup=PasswordForMUDBUser
#     OR
#   scriptName.sh --muDBPass=PasswordForMUDBUser
#
#
declare muDBPass="muDBPass"
#
# Comment out the next line if you want a simplified muDBUser password - both this and the initial/simple value above may be overridden with the CLI arguments
#
muDBPass=$(getRandPass 12 16)

#
# The mediaWikiPass value will be injected into the MediaWiki installation at runtime.
# 
# To set this value when invoking the script from commandline, call:
#   scriptName.sh -mwp=YourPass
#     OR
#   scriptname.sh --mediaWikiPass=YourPass
#
declare mediaWikiPass="mwikiRoot"
#
# Comment out the next line if you want a simplified Media Wiki Admin password - both this and the initial/simple value above may be overridden with the CLI arguments
#
mediaWikiPass=$(getRandPass 12 16)
#
# Process the CLI Parameters
#
for cliParam in "$@"; do
    #
    # $(cliParam,,) syntax converts the variable value to all lowercase
    # $(cliParam^^) syntax converts the variable value to all uppercase
    #
    #   * In this fashion parameter matching is case-insensitive
    #
    case ${cliParam,,} in
        -mrp=*|--mysqlrootpass=*)
            #
            # Need to pick parameter apart using string match specifications in your consuming function.
            #  ** Bash' native string matching/manipulation functions are quite simple:
            #      # removes the shortest match from the beginning of a string
            #      ## removes the longest match from the beginning
            #      % removes the shortest match from the end of a string
            #      %% removes the longest match from the end
            #  ***  As of 2017-08-27, https://spin.atomicobject.com/2014/02/16/bash-string-maniuplation/
            #       had some great information on Bash-native string manipulation
            #
            mysqlRootPass="${cliParam#*=}"
            shift 1
        ;;
        -mus=*|--mudbschema=*)
            muDBSchema="${cliParam#*=}"
            shift 1
        ;;
        -muu=*|--mudbuser=*)
            muDBUser="${cliParam#*=}"
            shift 1
        ;;
        -mup=*|--mudbpass=*)
            muDBPass="${cliParam#*=}"
            shift 1
        ;;
        -mwp=*|--mediawikipass=*)
            mediaWikiPass="${cliParam#*=}"
            shift 1
        ;;
        *)
            #
            # Default case is when the CLI parameter matches no defined script parameters.
            #
            # Shift CLI Parameters array left by 1 to eliminate this value
            #
            shift 1
            echo "Invalid command line parameter:  ${cliParam%%=*}"
            exit 1
        ;;
    esac
done

# testing OS (not processor) architecture
declare osArch=$(uname -p)
if [ "${osArch##*_}" == "64" ]
then
    sudo dpkg --add-architecture i386
fi

sudo apt-get --assume-yes update
sudo apt-get --assume-yes upgrade
# if you don't have these by default, WTF?!
sudo apt-get --assume-yes install build-essential libc6-dev lib32gcc1 lib32stdc++6
#
# the next round of installs are required for SteamCMD and some specific Steam games - they will probably be handy whether 
# you plan to host Steam games from your server or not
#
sudo apt-get --assume-yes install libvorbisfile3
#
# these are not typically installed by default on debian, maybe on some flavors of ubuntu, but are always useful
#
sudo apt-get --assume-yes install net-tools debconf-utils whois nmap inetutils-traceroute unzip expect curl
#
# if you're installing on a remote host and connected via SSH then, pretty obviously, you don't need the next line
# so you can safely comment it out.  you actively WANT to comment it out if you're using an SSH server other than openSSH,
# unless you want to install openSSH and remove whatever default you have installed
# if you're running debian or ubuntu then openssh is your default and the next line shouldn't hurt in either event
#
sudo apt-get --assume-yes install openssh-server openssh-client
#
# You need these for PennMUSH
#
sudo apt-get --assume-yes install openssl libssl-dev
#
# You need these for PennMUSH
#
sudo apt-get --assume-yes install libpcre3 libpcre3-dev libevent-dev
#
# Install email services
#   * You want sendmail for website fu - a LOT of pre-chewed website garbage you'll want to download and puke up on people uses it
#   * postfix is a dropin replacement for sendmail.  A LOT of discerning (smart, handsome, charming, talented) people prefer it over sendmail
#     for lots of reasons
#   * both postfix and sendmail are SMTP which is for MAIL-SENDING ONLY
#   * dovecot is for POP-3/IMAP which is for MAIL-RECEIVING
#
# sudo apt-get --assume-yes install sendmail-bin sendmail
sudo apt-get --assume-yes install postfix postfix-mysql dovecot-core dovecot-imapd dovecot-lmtpd dovecot-mysql
    # When Postfix configuration is prompted choose Internet Site:
    # Postfix configuration will ask about System mail name – you could use your FDQN or main domain.
#
# You want python with these addons at a bare minimum
#
sudo apt-get --assume-yes install python python-pip virtualenv
#
# Upgrade pip
#
sudo -H pip install --upgrade pip
#
# You want to be able to convert things to PDF on the server-side
#
sudo -H pip install pdfkit
sudo apt-get --assume-yes install wkhtmltopdf
#
# You want apache with these addons at a bare minimum
#
sudo apt-get --assume-yes install apache2 libapache2-mod-python libapache2-mod-php
#
# You want php with these addons at a bare minimum
#
sudo apt-get --assume-yes install php7.0 php-pear php7.0-mysql php7.0-curl php-xml
#
# You want git
#
sudo apt-get --assume-yes install git
#
# You'll want all the mysql libs to go along with these packages
#   * libmysqlclient-dev is default-libmysqlclient-dev under debian 9
#
sudo apt-get --assume-yes install mysql-client libmysqlclient-dev python-mysqldb 

#
# Apache2 config
#
sudo a2dismod mpm_event
sudo a2enmod mpm_prefork
sudo a2enmod ssl     # enable ssl security mod
sudo a2enmod headers # enable headers mod
sudo a2enmod rewrite # enable rewrite mod
sudo a2enmod cgi # enable cgi mod
sudo a2enmod python
sudo a2enmod php7.0
#
# PHP config
#
sudo phpenmod mbstring
sudo phpenmod xml

#
# Set up MySQL Server Password
#
debconf-set-selections <<< 'mariadb-server-5.5 mysql-server/root_password password ${mysqlRootPass}'
debconf-set-selections <<< 'mariadb-server-5.5 mysql-server/root_password_again password ${mysqlRootPass}'
# the next 2 lines will interrupt the hands-off installation process, prompting the user
# for a MySQL root password (twice) and then several more times
sudo apt-get --assume-yes install mysql-server


#
# script writer's recommendations for mysql_secure_installation's prompt answers:
#   First (obviously) enter your root password from when you installed mysql server
#   press N to NOT setup the VALIDATE PASSWORD plugin
#   press N to NOT change the root password (the one you already set is fine)
#   press Y to DO remove anonymous users
#   press Y to DO disallow root login remotely (you can only login from localhost, ie from the mysql app/shell, from PHP/Python/MUSH apps 
#     [not that you should ever log in as root from any app but the mysql shell app]) but the point is that the root username can only be used 
#     to log into mysql on the local system (yes you can still log in as root through SSH terminal connections - that's localhost as far as mysql 
#     is concerned - also be aware that mysql's root user and your linux box's root user are 2 different things)
#   press Y to DO remove test database and access to it (unless you're brand spanking new to MySQL and you really want or need the study material)
#   press Y to DO reload privilege tables and therefore make your changes effective as of now
#   * I'm pretty sure there's no platform where keystroke alone invokes the option - it's always Y or N followed by ENTER
#
# mysql_secure_installation

function doMySQLQuery
{
    expect -c "
    
    set timeout 10
    spawn mysql --user="root" -p -e \"$1\"

    expect \"Enter password: \"
    send \"${mysqlRootPass}\r\"

    expect eof
    "
}

echo -e '\033[1m\nRemoving root access from anywhere but localhost\n\033[0m'
doMySQLQuery "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1');"

echo -e '\033[1m\nDeleting anonymous user(s)\n\033[0m'
doMySQLQuery "DELETE FROM mysql.user WHERE User='';"

echo -e '\033[1m\nRemoving test/sample database\n\033[0m'
doMySQLQuery "DROP DATABASE IF EXISTS test;"

echo -e '\033[1m\nRemoving access to test/sample database\n\033[0m'
doMySQLQuery "DELETE FROM mysql.db WHERE Db='test' OR Db='test\\_%';"

echo -e '\033[1m\nUpdating privilege tables\n\033[0m'
doMySQLQuery "FLUSH PRIVILEGES;"

#
# Create directory for PHP logs
#
if [[ ! -d "/var/log/php" ]]; then
    sudo mkdir /var/log/php
    sudo chown www-data /var/log/php
fi

#
# Create the MU's MySQL Database
#
doMySQLQuery "CREATE DATABASE IF NOT EXISTS ${muDBSchema};"
doMySQLQuery "DROP USER IF EXISTS '${muDBUser}'@'localhost';"
doMySQLQuery "DROP USER IF EXISTS '${muDBUser}'@'127.0.0.1';"
doMySQLQuery "CREATE USER IF NOT EXISTS '${muDBUser}'@'localhost' IDENTIFIED BY '${muDBPass}';"
doMySQLQuery "CREATE USER IF NOT EXISTS '${muDBUser}'@'127.0.0.1' IDENTIFIED BY '${muDBPass}';"
doMySQLQuery "GRANT ALL PRIVILEGES ON `${muDBSchema}`.* TO '${muDBUser}'@'localhost';"
doMySQLQuery "GRANT ALL PRIVILEGES ON `${muDBSchema}`.* TO '${muDBUser}'@'127.0.0.1';"
doMySQLQuery "FLUSH PRIVILEGES;"

#
# These packages are required by MediaWiki
#
sudo apt-get --assume-yes install imagemagick php7.0-intl php7.0-gd php7.0-mbstring php-apcu
#
# This var contains a target URL for the current (or desired) release of mediawiki
#
declare mediaWikiURL="https://releases.wikimedia.org/mediawiki/1.30/mediawiki-1.30.0.tar.gz"
#
# The path to the local directory you intend to serve your MediaWiki site/content from
#
declare localWebSiteDir="/var/www/html/wiki"
if [[ ! -d "${localWebSiteDir}" ]]; then
    mkdir -p "${localWebSiteDir}"
    #
    # set website directory to RWX (OWNER) RX (GROUP) RX (OTHER)
    #
    chmod 755 "${localWebSiteDir}"
fi
#
# Download the target file to /opt/filename
#
sudo wget ${mediaWikiURL} --directory-prefix=/opt/
#
# Untar the target file into the website directory
#
sudo tar -xvzf /opt/${mediaWikiURL##*/} -C ${localWebSiteDir} --strip-components=1
#
# Set directory ownership for the mediawiki directory - don't ever modify this unless Apache decides to change its user/group from www-data
#
sudo chown www-data:www-data -R ${localWebSiteDir}
#
# Create databases for mediawiki
#
doMySQLQuery "SET GLOBAL sql_mode=''"
doMySQLQuery "CREATE DATABASE IF NOT EXISTS wikidb;"
doMySQLQuery "DROP USER IF EXISTS 'wikiadmin'@'localhost';"
doMySQLQuery "CREATE USER IF NOT EXISTS 'wikiadmin'@'localhost' IDENTIFIED BY '${mediaWikiPass}';"
doMySQLQuery "GRANT ALL PRIVILEGES ON wikidb.* TO 'wikiadmin'@'localhost';"
doMySQLQuery "FLUSH PRIVILEGES;"

# restart apache2
sudo systemctl restart apache2

#
# Set any other scripts in this directory executable - assuming you didn't jump the gun and do this manually, this is a sort of "enable the next step" operation 
# that helps alert you that it's now the appropriate time to do your MU installation
#
# These lines were commented out when it was discovered that some users (even highly experienced ones) were placing these scripts in bad locations - namely at the root of 
# their home directories - or that they were running this script repeatedly, even AFTER unpacking/installing MUSH/MUX platforms in subdirectories.  Bad mojo, y'all.
#
# So only uncomment these lines before running this script if you've quite cleverly placed this script in a CONTAINED directory such as /home/myUser/MUSH and it sits 
# alongside the other scripts that were distributed with this package, !!! AND NOTHING ELSE !!!
#
# FILES=$(find ./ -name '*.sh')
# for f in ${FILES}; do
#    chmod 700 ${f}
#    echo "set +x for ${f}"
# done

#
# There are certain things some hosts don't do for you.  We need to be sure those things are done for you.
#
if [[ ! -d "/etc/skel/.ssh" ]]; then
    sudo mkdir /etc/skel/.ssh
    sudo chmod 700 /etc/skel/.ssh
fi

if [[ ! -f "/etc/skel/.ssh/authorized_keys" ]]; then
    sudo touch /etc/skel/.ssh/authorized_keys
    sudo chmod 600 /etc/skel/.ssh/authorized_keys
fi

#
# Hopefully you were wise enough to copy the doBackups.sh script into the home directory where it belongs
#
sudo cp ~/doBackups.sh /etc/skel/doBackups.sh
sudo chmod 770 /etc/skel/doBackups.sh
#
# If you use nano, and I use nano, then you need line and column numbering and it's just plain stupid to have to toggle it.  Ever.
#
if [[ ! -f "~/.nanorc" ]]; then
    echo "set constantshow" > ~/.nanorc
fi
chmod 660 ~/.nanorc
#
# And do yourself a favor since, if you're brilliant, you're probably going to run multiple user accounts for yourself to use
#
sudo cp ~/.nanorc /etc/skel/.nanorc
sudo chmod 660 /etc/skel/.nanorc

echo ""
echo -e "\033[1m${BASH_SOURCE[0]##*/} Finished!\033[0m"
echo -e "\033[1m    Your MySQL/MariaDB root password is:  ${mysqlRootPass}\033[0m"
echo ""
echo -e "\033[1m    Your MU's DB Schema is:  ${muDBSchema}\033[0m"
echo -e "\033[1m    Your MU's DB User is:  ${muDBUser}\033[0m"
echo -e "\033[1m    Your MU's DB Password is:  ${muDBPass}\033[0m"
echo ""
echo -e "\033[1m    Your MediaWiki DB Schema is wikidb\033[0m"
echo -e "\033[1m    Your MediaWiki User is wikiadmin\033[0m"
echo -e "\033[1m    Your MediaWiki Password is ${mediaWikiPass}\033[0m'"
echo ""

#
# We do this LAST because it might disconnect SSH connections
#
# probably not needed in Ubuntu, but maybe debian and other flavors
#   * iptables is your firewall and ufw is "Uncomplicated Firewall" which is a simplification interface for iptables
#
sudo apt-get --assume-yes install iptables ufw

#
# Set up to deny all incoming connections that aren't explicitly allowed
#
sudo ufw default deny incoming
#
# Set up to allow all outgoing traffic
#
sudo ufw default allow outgoing

#
# Explicitly allow SSH
#
sudo ufw allow 22/tcp
#
# Allow HTTP and HTTPS
#
sudo ufw allow 80/tcp
sudo ufw allow 443/tcp

#
# You can also specify a specific port that the IP address is allowed to connect to by adding to any port followed by the port 
# number. For example, If you want to allow 15.15.15.51 to connect to port 22 (SSH), use this command:
#
# sudo ufw allow from 15.15.15.51 to any port 22
#

sudo ufw start

sudo ufw enable